### ISSUE TITLE
Potential mismatch in history window update logic (x_his / v_his) compared to Eq. (3)

### ISSUE THREAD

**Issue by Tang-Webber** [kywind/pgnd#9]
Hi, thanks for the great work!

I think there may be an off-by-one timing issue at train_eval.py L345:
https://github.com/kywind/pgnd/blob/ae050d1342faa0bceb2a10f4b0ab2e11682351cb/experiments/train/train_eval.py#L345

Tracing the execution order inside the rollout loop:

  L293  x_in = x.clone()                             # x_in = X_t (snapshot)
  L309  pred = material(x, v, x_his)           # model sees current = X_t 
  L310  x, v = sim(...)                                 # x is overwritten in place → X_{t+1}
  L345  x_his ← [x_his[:,1:], x[..].detach()]  # pushes X_{t+1}, not X_t

So the value appended to the history is X_{t+1} (post-sim), not X_t (pre-sim).
Could you confirm whether the current behavior is intentional, or if pushing x_in is the intended fix?

Thanks!

**Comment by kywind** [comment]
Hi, thanks for the feedback! Sorry for a few days late.

Yeah after re-checking everything, what you described does seem like an issue. It would probably also slightly affect training and evaluating the current checkpoints.

For now I made a fix in the history_window_fix branch: e6f535d749a42adb1708aa56a840247a3350f87e . When I finish re-training the models and verifying, I will merge it to the main branch.

Thank you for pointing this out! Let me know if you have other questions.

**Comment by kywind** [comment]
The fix has now been merged into main in 06ef5db building on the original fix in e6f535d. 

The rollout now snapshots x_in and v_in before simulation and appends those pre-simulation states to the history window afterward. This is applied consistently in training, evaluation, and real-world planning, including the gripper-point paths.

We also retrained and evaluated all categories on their complete held-out splits, with results broadly on par with the released checkpoints.

Thanks for the careful report! Closing this as resolved.
